jwst/linearity/tests/test_linearity.py

```diff
@@ -0,0 +1,596 @@
+from jwst.linearity.linearity import do_correction as lincorr
+from jwst.datamodels import dqflags, LinearityModel, MIRIRampModel
+from astropy.io import fits
+import numpy as np
+
+
+def test_coeff_dq():
+    """test linearity algorithm with random data ramp (does algorithm match expected algorithm)
+    also test a variety of dq flags and expected output """
+
+    # size of integration
+    nints = 1
+    ngroups = 160
+    xsize = 1032
+    ysize = 1024
+
+    # create the data and groupdq arrays
+    # create a JWST datamodel for MIRI data
+    im = make_rampmodel(nints, ngroups, ysize, xsize)
+    im.meta.instrument.detector = 'MIRIMAGE'
+
+    # Create reference file
+    numcoeffs = 5
+
+    ref_model = LinearityModel()
+    ref_model.data = np.zeros(shape=(numcoeffs, ysize, xsize), dtype=np.float32)
+    ref_model.dq = np.zeros((ysize, xsize), dtype=int)
+
+    ref_model.meta.instrument.name = 'MIRI'
+    ref_model.meta.instrument.detector = 'MIRIMAGE'
+    ref_model.meta.subarray.xstart = 1
+    ref_model.meta.subarray.ystart = 1
+    ref_model.meta.subarray.xsize = xsize
+    ref_model.meta.subarray.ysize = ysize
+
+    # Set coefficient values in reference file to check the algorithm
+    # Equation is DNcorr = L0 + L1*DN(i) + L2*DN(i)^2 + L3*DN(i)^3 + L4*DN(i)^4
+    # DN(i) = signal in pixel, Ln = coefficient from ref file
+    # L0 = 0 for all pixels for CDP6
+    L0 = 0
+    L1 = 0.85
+    L2 = 4.62E-6
+    L3 = -6.16E-11
+    L4 = 7.23E-16
+
+    coeffs = np.asfarray([0.0e+00,  0.85, 4.62e-06,  -6.16e-11, 7.23e-16])
+
+    ref_model.data[:, 300, 500] = coeffs
+    print(ref_model.data[:, 300, 500])
+
+    # check behavior with NaN coefficients: should not alter pixel values
+    coeffs2 = np.asfarray([L0, np.nan, L2, L3, L4])
+    ref_model.data[:, 200, 500] = coeffs2
+    im.data[0, 50, 200, 500] = 500.0
+
+    tgroup = 2.775
+
+    # set pixel values (DN) for specific pixels up the ramp
+    im.data[0, :, 300, 500] = np.arange(ngroups) * 100 * tgroup
+
+    scival = 40000.0
+    im.data[0, 45, 300, 500] = scival  # to check linearity multiplication is done correctly
+    im.data[0, 30, 350, 360] = 35  # pixel to check that dq=2 meant no correction was applied
+
+    # check if dq flags in pixeldq are correctly populated in output
+    im.pixeldq[50, 40] = 1
+    im.pixeldq[50, 41] = 2
+    im.pixeldq[50, 42] = 1024
+    im.pixeldq[50, 43] = 2048
+
+    # set dq flags in DQ of reference file
+    ref_model.dq[350, 350] = 1  # DO_NOT_USE
+    ref_model.dq[350, 360] = dqflags.pixel['NO_LIN_CORR']  # NO_LIN_CORR
+    ref_model.dq[300, 500] = 0  # good pixel
+
+    # run through Linearity pipeline
+    outfile = lincorr(im, ref_model)
+
+    # check that multiplication of polynomial was done correctly for specified pixel
+    outval = L0+(L1*scival)+(L2*scival**2)+(L3*scival**3)+(L4*scival**4)
+
+    print(outfile.data[0, 45, 300, 500])
+    assert(np.isclose(outfile.data[0, 45, 300, 500], outval, rtol=0.1))
```

Comment:
I think rtol=0.1 represents a relative tolerance of 10%, which is pretty big (if I'm correct). We should get matches down to at least ~5 significant digits (at 6 or 7 you could start running into machine precision issues, for single precision floats). So this could likely be changed to something on the order of rtol=0.00001.